src/Microsoft.OData.Core/ODataInnerError.cs

```diff
@@ -33,38 +41,71 @@ public ODataInnerError(Exception exception)
         {
             ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
-            this.Message = exception.Message ?? string.Empty;
-            this.TypeName = exception.GetType().FullName;
-            this.StackTrace = exception.StackTrace;
-
             if (exception.InnerException != null)
             {
                 this.InnerError = new ODataInnerError(exception.InnerException);
             }
+
+            Properties = new Dictionary<string, ODataValue>();
+
+            //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
+            ////       to stay compatible with Astoria.
+            Properties.Add(JsonConstants.ODataErrorInnerErrorMessageName, exception.Message.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorTypeNameName, exception.GetType().FullName.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorStackTraceName, exception.StackTrace.ToODataValue() ?? new ODataNullValue());
+        }
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="T:Microsoft.OData.ODataInnerError" /> class with a dictionary of property names and corresponding ODataValues.
+        /// </summary>
+        /// <param name="properties">Dictionary of string keys with ODataValue as value. Key string indicates the property name where as the value of the property is encapsulated in ODataValue.</param>
+        public ODataInnerError(IDictionary<string, ODataValue> properties)
+        {
+            ExceptionUtils.CheckArgumentNotNull(properties, "properties");
+
+            Properties = properties;
+        }
+
+        /// <summary>
+        /// Creates an instance of ODataInnerError with an empty property bag.
+        /// </summary>
+        /// <returns>returns an instance of ODataInnerError with empty property bag.</returns>
+        public static ODataInnerError CreateEmptyInnerError()
+        {
+            return new ODataInnerError(new Dictionary<string, ODataValue>());
+        }
+
+        /// <summary>
+        /// Properties to be written for the inner error.
+        /// </summary>
+        public IDictionary<string, ODataValue> Properties

```

Comment:
>Properties [](start = 47, length = 10)

The code doesn't seem to support properties of type ODataCollectionValue. If this is a restriction, we might want to wrap the IDictionary and throw an error if the user tries to set a value that is an ODataCollectionValue.  Might be easier just to support reading/writing the collection value...